src/js/components/stock-movement-wizard/outbound/AddItemsPage.jsx

```diff
@@ -709,49 +735,97 @@ class AddItemsPage extends Component {
-                _.includes(savedItemsProductCodes, item.product.productCode)
+                _.includes(savedItemsProductCodes, item.product?.productCode)
                 && parseInt(item.quantityRequested, 10) > 0
+                && item.status === SaveStatus.PENDING
               ) {
-                return { ...item, disabled: true };
+                return { ...itemToChange, disabled: true, status: SaveStatus.SAVED };
               }
-
               return item;
             });
-            this.setState({ values: { ...this.state.values, lineItems: lineItemsAfterSave } });
+            this.setState({
+              values: { ...this.state.values, lineItems: lineItemsAfterSave },
+              currentLineItems: lineItemsBackendData,
+            });
             return;
           }
 
-          this.setState({ values: { ...this.state.values, lineItems: lineItemsBackendData } });
-
           this.setState({
+            values: { ...this.state.values, lineItems: lineItemsBackendData },
             currentLineItems: lineItemsBackendData,
           });
         })
-        .catch(() => Promise.reject(new Error(this.props.translate('react.stockMovement.error.saveRequisitionItems.label', 'Could not save requisition items'))));
+        .catch(() => {
+          // When there is an error during saving we have to find products which
+          // caused the error. These items are not saved, so we don't have line ID,
+          // and we have to find these items by product ID and SaveStatus
+          const notSavedItemsIds = payload.lineItems.map(item => item['product.id']);
+          const lineItemsWithErrors = this.state.values.lineItems.map((item) => {
+            if (
+              item.product &&
+              item.status === SaveStatus.PENDING &&
+              _.includes(notSavedItemsIds, item.product.id)
+            ) {
+              return { ...item, status: SaveStatus.ERROR };
+            }
+            return item;
+          });
+          this.setState({ values: { ...this.state.values, lineItems: lineItemsWithErrors } });
+          return Promise.reject(new Error(this.props.translate('react.stockMovement.error.saveRequisitionItems.label', 'Could not save requisition items')));
+        });
     }
 
     return Promise.resolve();
   }
 
-  saveProgress = (values) => {
-    this.setState({ isSaveCompleted: false });
+  // if rowIndex is passed, it means that we are editing row
+  // not adding new one
+  saveProgress = ({ values, rowIndex, fieldValue }) => {
+    // I can't check !!rowIndex, because it can be 0,
+    // so there is possibility to return false when the
+    // rowIndex is present
+    const isEdited = rowIndex !== undefined;
+    const itemsWithStatuses = values.lineItems.map((item) => {
+      if (isEdited && rowIndex === values.lineItems.indexOf(item)) {
+        return { ...fieldValue, status: SaveStatus.PENDING };
+      }
+
+      if (item.product && parseInt(item.quantityRequested, 10) <= 0) {
```

Comment:
is `quantityRequested` the only field that is validated in this workflow?

Reply:
There are 3 fields: requested product, quantity and recipient. If we don't have product and quantity, the line will be `pending`. Recipient is not necessary, so there is no need for validating this field. Correct me if I am wrong.

Reply:
this is what I remembered, but got confused about the `validate` function which also contains checking for `item.boxName` and `item.palletName`. 

By the way thing that came to my mind now is that maybe instead of checking it manually here and in the `validate` function, we could here maybe just call the `validate` function and just check if returned errors are an empty object? 
What do you think? We'd obviously have to refactor `validate` function a bit or just call it with a proper argument (not item, because it expects something like:
```
{ 
  values: {
    lineItems: [item]
  }
}
```

but by doing so, if we wanted to refactor/add/remove any validation, we wouldn't have to do it in a few places, but just one.

Reply:
Validating function checks if field is empty or if there is a wrong value. If the field is empty, line should have `pending` status, so if i would like to use this function I have to check which type of error it is, so i think there will be some additional unnecessary code. I just need only one exact error from the validating function, so I think this will be overcomplicated if I use it.

Reply:
Ok, I get the point now